peps/pep-0733.rst

```diff
@@ -0,0 +1,629 @@
+add a concrete type alternative. The principle around this should
+be documented in the guidelines
+[`Issue 23 <https://github.com/capi-workgroup/problems/issues/23>`__].
+
+Implementation Flaws
+--------------------
+
+Below is a list of localized implementation flaws. Most of these can
+probably be fixed incrementally, if we choose to do so. They should,
+in any case, be avoided in any new API design.
+
+There are functions that don't follow the convention of
+returning ``0`` for success and ``-1`` for failure. For
+example, ``PyArg_ParseTuple`` returns 0 for success and
+non-zero for failure
+[`Issue 25 <https://github.com/capi-workgroup/problems/issues/25>`__].
+
+The macros ``Py_CLEAR`` and ``Py_SETREF`` access their arg more than
+once, so if the arg is an expression with side effects, they are
+duplicated
+[`Issue 3 <https://github.com/capi-workgroup/problems/issues/3>`__].
+
+The meaning of ``Py_SIZE`` depends on the type and is not always
+reliable
+[`Issue 10 <https://github.com/capi-workgroup/problems/issues/10>`__].
+
+Some API function do not have the same behaviour as their Python
+equivalents.  The behaviour of ``PyIter_Next`` is different from
+``tp_iternext``.
+[`Issue 29 <https://github.com/capi-workgroup/problems/issues/29>`__].
+The behaviour of ``PySet_Contains`` is different from ``set.__contains__``
+[`Issue 6 <https://github.com/capi-workgroup/problems/issues/6>`__].
+
+The fact that ``PyArg_ParseTupleAndKeywords`` takes a non-const
+``char*`` array as argument makes it more difficult to use
+[`Issue 28 <https://github.com/capi-workgroup/problems/issues/28>`__].
+
+``Python.h`` does not expose the whole API. Some headers (like ``marshal.h``)
+are not included from ``Python.h``.
+[`Issue 43 <https://github.com/capi-workgroup/problems/issues/43>`__].
+
+**Naming**
+
+``PyLong`` and ``PyUnicode`` use names which don't match the Python
+types they represent (``int``/``str``). This can be fixed in a new API
+[`Issue 14 <https://github.com/capi-workgroup/problems/issues/14>`__].
+
+There are identifiers in the API which are lacking a ``Py``/``_Py``
+prefix
+[`Issue 46 <https://github.com/capi-workgroup/problems/issues/46>`__].
+
+Missing Functionality
+---------------------
+
+This section consists of a list of feature requests, i.e., functionality
+that was identified as missing in the current C API.
+
+Debug Mode
+~~~~~~~~~~
+
+A debug mode that can be activated without recompilation and which
+activates various checks that can help detect various types of errors
+[`Issue 36 <https://github.com/capi-workgroup/problems/issues/36>`__].
+
+Introspection
+~~~~~~~~~~~~~
+
+There aren't currently reliable introspection capabilities for objects
+defined in C in the same way as there are for Python objects
+[`Issue 32 <https://github.com/capi-workgroup/problems/issues/32>`__].
+
+Efficient type checking for heap types, similar to what ``Py*_Check``
+can do for a static type
```

Comment:
Used to be able to do - they're equally slow now, right? Only efficient when we know that subclassing isn't possible.

Reply:
Last I looked, the C-API type checks for builtin types were actually efficiently using type flags.

Reply:
"Static" is not the same as "builtin". The ones with flags are more special than static types.

Maybe built-in was intended here? I didn't want to presume that. But it's very unlikely we'll be able to give all types a flag that is inherited by subclasses, so if that's the intention, I'd vote to rule out the suggestion now.

Reply:
Yeah, they still can, but arbitrary static types could not be checked using a bitmask - only those that inherited from a builtin type.

Reply:
Ah, right. Well, in any case, efficient type checking is important for several applications, to the point that I committed [this horror](https://github.com/apache/arrow/blob/d68f8e21643e8aec9ed253094edfd15c0a08f1c1/python/pyarrow/src/arrow/python/helpers.cc#L277-L284) at some point.

Reply:
I don't disagree, but the cases here are dictated by Python's semantics and not by the C API:
* you have a `type` object and an instance of that type (fast)
* you have a `type` object and an instance of a subclass of that type (slow)
* you have a `type` object and an instance of a type that implements `__instancecheck__` or `__subclasscheck__` (slower)
* you don't have a `type` object (umm... not sure how to generalise type checking when you don't have a type)

Boolean flags are an optimisation for certain built-in types only, we can't use them as the expectation for all types. The general case (i.e. for anything not built into the runtime) is a C version of `isinstance(o, T)` or `type(o) is T`.

Reply:
Well, whether the general case can be made fast is an open question, though there is no reason that it couldn't.

Reply:
There are also two kinds of type checking: `PyObject_IsSubclass` (which allows `__*check__` hooks) and `PyType_IsSubtype` (which should guarantee compatible memory layout, i.e. whether it's safe to cast).
In the C-API you usually care about the latter. We could allow checking compatible memory layout without an actual `type` object to compare to (see [#17](https://github.com/capi-workgroup/problems/issues/17)), and you could say that's just a workaround for an even thornier problem -- but IMO, that level of detail is out of scope for this PEP.

Reply:
Thanks all. I'll just remove the contentious part of this sentence, we can do without it.